### ISSUE TITLE
Hide all requests from a user and ban them from the admin interface

### ISSUE THREAD

**Issue by garethrees** [mysociety/alaveteli#3580]
Requires https://github.com/mysociety/alaveteli/issues/3578

**Comment by RichardTaylor** [comment]
Just adding a +1 here. On WhatDoTheyKnow we've had a banned user creating new accounts to evade a ban and they've been making 2-4 requests plus annotations on the accounts. Hiding the requests and annotations at the moment is taking quite a number of clicks.

**Comment by garethrees** [comment]
⚠️ **EDIT:** Updated in https://github.com/mysociety/alaveteli/issues/3580#issuecomment-1049996116

Here's a bit of code to do this through the console:

```ruby
def hide_all_requests(user_id, admin_url_name)
  InfoRequest.where(user_id: user_id).find_each do |info_request|
    old_prominence = info_request.prominence

    next if old_prominence == 'hidden'

    info_request.update!(prominence: 'hidden')

    params =
      { old_prominence: old_prominence,
        prominence: info_request.prominence,
        editor: admin_url_name }

    info_request.log_event('edit', params)
  end
end

# Example:
#
# hide_all_requests(181416, 'gareth_rees_3')

```

Can obviously tweak the prominence to whatever final state is most relevant.

**Comment by itsaphel** [comment]
Would it be safe to automatically do this for any "Ban for spamming" ban, rather than make it a separate function?

**Comment by garethrees** [comment]
It's very unusual for us to use the "Ban for spamming" button for human accounts; it's more for automated signups from bots.

I don't think I'd want to connect something like this to an account suspension action because it reduces our flexibility in decision-making. For example, we might decide to hide all a user's existing requests and send a warning not to continue as they were before. We wouldn't _necessarily_ also ban them (despite what the issue title says). It does add a little extra work for admins, (update ban text; then click hide all requests) but for that we gain optionality, which is good.

I think the _ideal_ UI for this would be similar to the comments (https://github.com/mysociety/alaveteli/pull/1398), where we can select a number of requests, and then have a variety of buttons (`hide all`, `backpage all`, etc) that submits to, say, `Admin::InfoRequestsBulkProminenceController#update` with the IDs of the requests and a param for the `prominence` value. We'd also want to add a javascript-based "Select All" link (a la https://github.com/mysociety/alaveteli/issues/3324).

<img width="1037" alt="Screenshot 2021-07-28 at 09 30 08" src="https://user-images.githubusercontent.com/282788/127290821-20c3ee6a-5865-4072-b8b2-7ba8fcd501a9.png">

That's quite a bit of work, so I think we could simplify by implementing the controller as described, and then just adding a variety of buttons ("Backpage All", "Requester Only All", "Hide all") that set up the params along the lines of:

```erb
<%= form_for … %>
  <%= hidden_field_tag :request_ids, @admin_user.info_requests.pluck(:id) %>
  <%= hidden_field_tag :prominence, 'hidden' %>
  <%= submit_tag 'Hide all requests' %>
<% end %>

<%= form_for … %>
  <%= hidden_field_tag :request_ids, @admin_user.info_requests.pluck(:id) %>
  <%= hidden_field_tag :prominence, 'backpage' %>
  <%= submit_tag 'Backpage all requests' %>
<% end %>

<!-- etc -->
```

I think we'd want a model-layer method of `InfoRequest#set_prominence_with_event` and pass the prominence value (`:hidden`, `:backpage`, `:requester_only`), `prominence_reason` ("Bulk prominence update") and the current admin user.

**Comment by itsaphel** [comment]
Thanks @garethrees! I think I understand the use case better now too. Will give that implementation you mention a shot at some point.

**Comment by garethrees** [comment]
No problem, and thank _you_ for looking in to this! 🙇  Feel free to post a spike if you want any early feedback.

**Comment by garethrees** [comment]
Can now also hide comments programatically from the console or app https://github.com/mysociety/alaveteli/issues/6812.

**Comment by garethrees** [comment]
⚠️ **EDIT:** Specific comments issue (with updated code) in https://github.com/mysociety/alaveteli/issues/6812.

Bit of an update on https://github.com/mysociety/alaveteli/issues/3580#issuecomment-557102002 (minor tweaks to `hide_all_requests` and incorporating comments:

```ruby
def hide_all_requests(user, editor)
  InfoRequest.where(user_id: user.id).find_each do |info_request|
    old_prominence = info_request.prominence

    next if %w[hidden requester_only].include?(old_prominence)

    info_request.update!(prominence: 'hidden')

    params =
      { old_prominence: old_prominence,
        prominence: info_request.prominence,
        editor: editor.url_name }

    info_request.log_event('hide', params)
  end
end

# def hide_all_comments(user, editor)
#   user.comments.find_each do |comment|
#     next if comment.hidden?
#     comment.hide(editor: editor)
#     comment.update(attention_requested: false)
#   end
# end

editor = User.find(USER_ID)

bad_users =
  %w[foo@example.com
     bar@example.com]

bad_users.map { |email| User.find_user_by_email(email) }.each do |user|
  hide_all_requests(user, editor)
#   hide_all_comments(user, editor)
end


```

**Comment by itsaphel** [comment]
@garethrees if I understand right, there's currently no pagination for comments/annotations? If pagination was added in the future, then I imagine the JS 'select all' approach might not work?

**Comment by garethrees** [comment]
> if I understand right, there's currently no pagination for comments/annotations?

That's correct, though it's been suggested as a solution to https://github.com/mysociety/alaveteli/issues/6980.

>  If pagination was added in the future, then I imagine the JS 'select all' approach might not work?

In this situation I'd imagine the process being:

* Select all
* Click appropriate hide button
* Go to next page
* Select all
* Click appropriate hide button
* and so on…

If we implemented the select boxes for _requests_, we'd probably then phrase the buttons "Hide _selected_ requests" rather than "Hide _all_ requests" etc.

**Comment by garethrees** [comment]
Linking to https://github.com/mysociety/alaveteli/issues/843, since more use of this would lead to more support requests for copies of the requests.

**Comment by garethrees** [comment]
This is desirable, but unlikely to be worked on in the next 12 months so closing for now.
